python/ray/_private/accelerators/neuron.py

```diff
@@ -0,0 +1,125 @@
+import os
+import sys
+import json
+import logging
+import subprocess
+from typing import Optional, List
+
+from ray._private.accelerators.accelerator import Accelerator
+
+logger = logging.getLogger(__name__)
+
+NEURON_RT_VISIBLE_CORES_ENV_VAR = "NEURON_RT_VISIBLE_CORES"
+NOSET_AWS_NEURON_RT_VISIBLE_CORES_ENV_VAR = (
+    "RAY_EXPERIMENTAL_NOSET_NEURON_RT_VISIBLE_CORES"
+)
+
+# https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-hardware/inf2-arch.html#aws-inf2-arch
+# https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-hardware/trn1-arch.html#aws-trn1-arch
+# Subject to removal after the information is available via public API
+AWS_NEURON_INSTANCE_MAP = {
+    "trn1.2xlarge": 2,
+    "trn1.32xlarge": 32,
+    "trn1n.32xlarge": 32,
+    "inf2.xlarge": 2,
+    "inf2.8xlarge": 2,
+    "inf2.24xlarge": 12,
+    "inf2.48xlarge": 24,
+}
+
+
+class NeuronAccelerator(Accelerator):
+    """AWS Inferentia and Trainium accelerators."""
+
+    @staticmethod
+    def get_resource_name() -> str:
+        return "neuron_cores"
+
+    @staticmethod
+    def is_available() -> bool:
+        return NeuronAccelerator.get_num_accelerators() > 0
+
+    @staticmethod
+    def get_visible_accelerator_ids_env_var() -> str:
+        return NEURON_RT_VISIBLE_CORES_ENV_VAR
+
+    @staticmethod
+    def get_visible_accelerator_ids() -> Optional[List[str]]:
+        neuron_visible_cores = os.environ.get(NEURON_RT_VISIBLE_CORES_ENV_VAR, None)
+
+        if neuron_visible_cores is None:
+            return None
+
+        if neuron_visible_cores == "":
+            return []
+
+        return list(neuron_visible_cores.split(","))
+
+    @staticmethod
+    def get_num_accelerators() -> int:
+        """
+        Attempt to detect the number of Neuron cores on this machine.
+
+        Returns:
+            The number of Neuron cores if any were detected, otherwise 0.
+        """
+        nc_count: int = 0
+        neuron_path = "/opt/aws/neuron/bin/"
```

Comment:
Note: The path doesn't guarantee for all environments (based on recent slack conversation https://ray-distributed.slack.com/archives/C01DLHZHRBJ/p1696448156026509)

Expected:
 find (neuron-ls) command if it exists
 then run the command to get the cores

Also, we did ask AWS Neuron SDK team to get the cores information using IMDS (API driven) but no plans to support this.

Reply:
I just copy pasted from the existing implementation. Are you planning to fix it for other environments?

Reply:
Can we add a try/catch and fix it in the current PR?
If not, ok to move it to issue and I'll own it (tentative ETA: Q1 2024).

Reply:
Also add TODO maybe?